pkg/strategy/xmaker/hedgemarket.go

```diff
@@ -0,0 +1,285 @@
+
+	tradeCollector := core.NewTradeCollector(symbol, position, orderStore)
+	tradeCollector.BindStream(session.UserDataStream)
+
+	activeMakerOrders := bbgo.NewActiveOrderBook(symbol)
+	activeMakerOrders.BindStream(session.UserDataStream)
+
+	logger := log.WithFields(logrus.Fields{
+		"exchange":     session.ExchangeName,
+		"hedge_market": market.Symbol,
+	})
+
+	m := &HedgeMarket{
+		session:   session,
+		market:    market,
+		symbol:    market.Symbol,
+		stream:    stream,
+		book:      book,
+		depthBook: depthBook,
+
+		connectivity: connectivity,
+
+		positionExposure: newPositionExposure(symbol),
+
+		positionDeltaC:    make(chan fixedpoint.Value, 5),
+		position:          position,
+		orderStore:        orderStore,
+		tradeCollector:    tradeCollector,
+		activeMakerOrders: activeMakerOrders,
+
+		logger: logger,
+	}
+
+	tradeCollector.OnTrade(func(trade types.Trade, _, _ fixedpoint.Value) {
+		delta := trade.PositionDelta()
+		m.positionExposure.Close(delta)
+
+		m.logger.Infof("trade collector received trade: %+v, position delta: %f, covered position: %f",
+			trade, delta.Float64(), m.positionExposure.pending.Get().Float64())
+
+		// TODO: pass Environment to HedgeMarket
+		/*
+			if profit.Compare(fixedpoint.Zero) == 0 {
+				s.Environment.RecordPosition(s.Position, trade, nil)
+			}
+		*/
+	})
+
+	return m
+}
+
+func (m *HedgeMarket) newMockTrade(
+	side types.SideType, price, quantity fixedpoint.Value, tradeTime time.Time,
+) types.Trade {
+	m.mockTradeId++
+
+	return types.Trade{
+		ID:            m.mockTradeId,
+		OrderID:       m.mockTradeId,
+		Exchange:      m.session.ExchangeName,
+		Price:         price,
+		Quantity:      quantity,
+		QuoteQuantity: quantity.Mul(price),
+		Symbol:        m.symbol,
+		Side:          side,
+		IsBuyer:       side == types.SideTypeBuy,
+		IsMaker:       true,
+		Time:          types.Time(tradeTime),
+		Fee:           fixedpoint.Zero,
+		FeeCurrency:   "",
+		StrategyID: sql.NullString{
+			String: ID,
+			Valid:  true,
+		},
+	}
+
+}
+
+func (m *HedgeMarket) submitOrder(ctx context.Context, submitOrder types.SubmitOrder) (*types.Order, error) {
+	submitOrder.Market = m.market
+	submitOrder.Symbol = m.symbol
+
+	submitOrders := []types.SubmitOrder{
+		submitOrder,
+	}
+
+	orderCreateCallback := func(createdOrder types.Order) {
+		m.orderStore.Add(createdOrder)
+
+		// for track non-market-orders
+		if createdOrder.Type != types.OrderTypeMarket {
+			m.activeMakerOrders.Add(createdOrder)
+		}
+	}
+
+	defer m.tradeCollector.Process()
+
+	createdOrders, _, err := bbgo.BatchPlaceOrder(
+		ctx, m.session.Exchange, orderCreateCallback, submitOrders...,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to submit order: %w", err)
+	}
+
+	if len(createdOrders) == 0 {
+		return nil, fmt.Errorf("no hedge order created")
+	}
+
+	createdOrder := createdOrders[0]
+	return &createdOrder, nil
+}
+
+func (m *HedgeMarket) getQuotePrice() (bid, ask fixedpoint.Value) {
```

Comment:
In most cases, hedging requires only the price of one side. How about adding an input flag to decide whether bid or ask price should be returned?

Reply:
I will add this